File: 130-surrounded-regions/surrounded-regions.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        neighbours = [(0,1),(0,-1),(1,0),(-1,0)]
        seen = set()
        rows = len(board)
        cols = len(board[0])

        def dfs(row: int, col: int) -> None:
            stack = [(row,col)]
            check = [(row,col)]
            island = True

            while check:
                row,col = check.pop()

                for n in neighbours:
                    if 0 <= row+n[0] < rows and 0 <= col+n[1] < cols:
                        if board[row+n[0]][col+n[1]] == "O" and (row+n[0],col+n[1]) not in seen:
                            seen.add(((row+n[0],col+n[1])))
                            stack.append((row+n[0],col+n[1]))
                            if 0 < row+n[0] < rows-1 and 0 < col+n[1] < cols-1:
                                check.append((row+n[0],col+n[1]))
                            else:
                                island = False
                        if (row+n[0],col+n[1]) in seen and ((row+n[0] == 0 or row+n[0] == rows-1) or (col+n[1] == 0 or col+n[1] == cols-1)):
                            island = False
            if island:
                for row,col in stack:
                    board[row][col] = "X"
    


        for i in range(rows):
            for j in range(cols):
                if board[i][j] == "O" and (i,j) not in seen:
                    seen.add((i,j))
                    if 0 < i < rows-1 and 0 < j < cols-1:
                        dfs(i,j)

        

File: 130-surrounded-regions/surrounded-regions.java
class Solution {
    private int[][] neighbours = {{1,0},{-1,0},{0,1},{0,-1}};
    private int rows;
    private int cols;


    public void solve(char[][] board) {
        rows = board.length;
        cols = board[0].length;

        for (int i = 0; i < rows; i++){
            if (board[i][0] == 'O') dfs(i,0,board);
            if (board[i][cols-1] == 'O') dfs(i,cols-1,board);
        }

        for (int i = 0; i < cols; i++){
            if (board[0][i] == 'O') dfs(0,i,board);
            if (board[rows-1][i] == 'O') dfs(rows-1,i,board);
        }

        for (int i = 0; i < rows; i++){
            for (int j = 0; j < cols; j++){
                if (board[i][j] == 'O') board[i][j] = 'X';
                if (board[i][j] == 'T') board[i][j] = 'O';
            }
        }
        
    }

    private void dfs(int row, int col, char[][] board){
        board[row][col] = 'T';

        for (int[] n: neighbours){
            int newrow = row+ n[0];
            int newcol = col + n[1];

            if (newrow < 0 || newrow >= rows || newcol < 0 || newcol >= cols ||
            board[newrow][newcol] != 'O') continue;

            dfs(newrow,newcol,board);
        }
            
      
    }
}